client/webserver/webserver.go

```diff
@@ -0,0 +1,401 @@
+// This code is available on the terms of the project LICENSE.md file,
+// also available online at https://blueoakcouncil.org/license/1.0.0.
+
+package webserver
+
+import (
+	"context"
+	"encoding/hex"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"math/rand"
+	"net"
+	"net/http"
+	"os"
+	"path"
+	"path/filepath"
+	"strings"
+	"sync"
+	"time"
+
+	"decred.org/dcrdex/client/core"
+	"github.com/decred/slog"
+	"github.com/go-chi/chi"
+	"github.com/go-chi/chi/middleware"
+	"github.com/gorilla/websocket"
+)
+
+const (
+	// rpcTimeoutSeconds is the number of seconds a connection to the
+	// RPC server is allowed to stay open without authenticating before it
+	// is closed.
+	rpcTimeoutSeconds = 10
+	// darkModeCK is the cookie key for dark mode.
+	darkModeCK = "darkMode"
+	// authCK is the authorization token cookie key.
+	authCK = "dexauth"
+	// authCV is the authorization middleware contextual value key.
+	authCV = "authctx"
+)
+
+var (
+	log slog.Logger
+	// websocket.Upgrader is the preferred method of upgrading a request to a
+	// websocket connection.
+	upgrader = websocket.Upgrader{}
+)
+
+// clientCore is satisfied by core.Core.
+type clientCore interface {
+	ListMarkets() []*core.MarketInfo
+	Register(*core.Registration) error
+	Login(dex, pw string) error
+	Sync(dex, mkt string) (*core.OrderBook, chan *core.BookUpdate, error)
+	Unsync(dex, mkt string)
+	Balance(string) (float64, error)
+}
+
+// marketSyncer is used internally to synchronize market subscriptions.
+type marketSyncer struct {
+	market string
+	kill   func()
+}
+
+// webServer is a single-client http and websocket server enabling a browser
+// interface to the DEX client.
+type webServer struct {
+	ctx        context.Context
+	core       clientCore
+	html       *templates
+	indent     bool
+	mtx        sync.RWMutex
+	authToken  string
+	openMarket *marketSyncer
+}
+
+// Run is the single exported function of webserver. Run the server, listening
+// on the provided network address until the context is canceled.
+func Run(ctx context.Context, core clientCore, addr string, logger slog.Logger, reloadHTML bool) {
```

Comment:
If we accept the new pattern of constructors making types with Run(ctx) methods, we'll probably follow up by refactoring the context passing in webServer and maybe in dex/ws.WSLink, but that's up in the air so let's run with this approach for now.